Unable to exclude a list of embedded entities from indexes using the `@Unindexed` annotation

When persisting an entity with a field containing a list of embedded entities, it is not possible to apply the `@Unindexed` annotation. When doing so, the following exception is thrown:

```
...
Caused by: com.google.datastore.v1.client.DatastoreException: Exclude from indexes cannot be set on a list value
...
```

Example entity:

```
@Entity
public class Entity {
    @Id private String id;
    @Unindexed private List<EmbeddedEntity> embeddedEntities;

    public static class EmbeddedEntity {
        private String name;
    }
}
```

I think it would be useful to have the library detect when the annotation is applied to a list value and apply the exclude from indexes property as appropriate to each of the list value's items.

I can achieve what I want by applying the `@Unindexed` annotation to each of the embedded entity's fields, however, this can be a little cumbersome.

--- a/docs/src/main/asciidoc/datastore.adoc
+++ b/docs/src/main/asciidoc/datastore.adoc
@@ -772,6 +772,9 @@ public class ExampleItem {
 
 	@Unindexed
 	long unindexedField;
+
+	@Unindexed
+	List<String> unindexedListField;
 }
 ----
 

--- a/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/DefaultDatastoreEntityConverter.java
+++ b/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/DefaultDatastoreEntityConverter.java
@@ -24,9 +24,9 @@ import java.util.Set;
 import java.util.stream.Collectors;
 
 import com.google.cloud.datastore.BaseEntity;
+import com.google.cloud.datastore.ListValue;
 import com.google.cloud.datastore.StringValue;
 import com.google.cloud.datastore.Value;
-import com.google.cloud.datastore.ValueBuilder;
 
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreMappingContext;
@@ -180,9 +180,7 @@ public class DefaultDatastoreEntityConverter implements DatastoreEntityConverter
 						Value convertedVal = this.conversions.convertOnWrite(val, persistentProperty);
 
 						if (persistentProperty.isUnindexed()) {
-							ValueBuilder valueBuilder = convertedVal.toBuilder();
-							valueBuilder.setExcludeFromIndexes(true);
-							convertedVal = valueBuilder.build();
+							convertedVal = setExcludeFromIndexes(convertedVal);
 						}
 						sink.set(persistentProperty.getFieldName(), convertedVal);
 					}
@@ -193,5 +191,17 @@ public class DefaultDatastoreEntityConverter implements DatastoreEntityConverter
 								ex);
 					}
 				});
+	}
+
+	private Value setExcludeFromIndexes(Value convertedVal) {
+		// ListValues must have its contents individually excluded instead.
+		// the entire list must NOT be excluded or there will be an exception.
+		if (convertedVal.getClass().equals(ListValue.class)) {
+			return ListValue.of(((ListValue) convertedVal).get().stream()
+					.map(val -> val.toBuilder().setExcludeFromIndexes(true).build()).collect(Collectors.toList()));
+		}
+		else {
+			return convertedVal.toBuilder().setExcludeFromIndexes(true).build();
+		}
 	}
 }

--- a/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/core/convert/DefaultDatastoreEntityConverterTests.java
+++ b/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/core/convert/DefaultDatastoreEntityConverterTests.java
@@ -607,6 +607,7 @@ public class DefaultDatastoreEntityConverterTests {
 		UnindexedTestDatastoreItem item = new UnindexedTestDatastoreItem();
 		item.setIndexedField(1L);
 		item.setUnindexedField(2L);
+		item.setUnindexedStringListField(Arrays.asList("a", "b"));
 
 		Entity.Builder builder = getEntityBuilder();
 		ENTITY_CONVERTER.write(item, builder);
@@ -622,6 +623,14 @@ public class DefaultDatastoreEntityConverterTests {
 				.as("validate excludeFromIndexes on indexed field").isFalse();
 		assertThat(entity.getValue("unindexedField").excludeFromIndexes())
 				.as("validate excludeFromIndexes on unindexed field").isTrue();
+
+		// the ListValue itself must NOT be unindexed or it will cause exception. the contents
+		// individually must be.
+		assertThat(entity.getValue("unindexedStringListField").excludeFromIndexes()).isFalse();
+		assertThat(((ListValue) entity.getValue("unindexedStringListField")).get().get(0).excludeFromIndexes())
+				.isTrue();
+		assertThat(((ListValue) entity.getValue("unindexedStringListField")).get().get(1).excludeFromIndexes())
+				.isTrue();
 	}
 
 	@Test

--- a/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/core/convert/UnindexedTestDatastoreItem.java
+++ b/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/core/convert/UnindexedTestDatastoreItem.java
@@ -16,18 +16,24 @@
 
 package org.springframework.cloud.gcp.data.datastore.core.convert;
 
+import java.util.List;
+
 import org.springframework.cloud.gcp.data.datastore.core.mapping.Unindexed;
 
 /**
- * A test entity with an unindexed property.
+ * A test entity with an unindexed property and unindexed list property.
  *
  * @author Dmitry Solomakha
+ * @author Chengyuan Zhao
  */
 public class UnindexedTestDatastoreItem {
 	private long indexedField;
 
 	@Unindexed
 	private long unindexedField;
+
+	@Unindexed
+	private List<String> unindexedStringListField;
 
 	public long getIndexedField() {
 		return this.indexedField;
@@ -43,5 +49,13 @@ public class UnindexedTestDatastoreItem {
 
 	public void setUnindexedField(long unindexedField) {
 		this.unindexedField = unindexedField;
+	}
+
+	public List<String> getUnindexedStringListField() {
+		return this.unindexedStringListField;
+	}
+
+	public void setUnindexedStringListField(List<String> unindexedStringListField) {
+		this.unindexedStringListField = unindexedStringListField;
 	}
 }

--- a/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/it/DatastoreIntegrationTests.java
+++ b/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/it/DatastoreIntegrationTests.java
@@ -43,6 +43,7 @@ import org.springframework.cloud.gcp.data.datastore.core.mapping.Descendants;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DiscriminatorField;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.DiscriminatorValue;
 import org.springframework.cloud.gcp.data.datastore.core.mapping.Entity;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.Unindexed;
 import org.springframework.cloud.gcp.data.datastore.it.TestEntity.Shape;
 import org.springframework.cloud.gcp.data.datastore.repository.DatastoreRepository;
 import org.springframework.cloud.gcp.data.datastore.repository.query.Query;
@@ -550,6 +551,8 @@ public class DatastoreIntegrationTests extends AbstractDatastoreIntegrationTests
 		SubEntity subEntity3 = new SubEntity();
 		SubEntity subEntity4 = new SubEntity();
 
+		subEntity1.stringList = Arrays.asList("a", "b");
+
 		ParentEntity parentEntity = new ParentEntity(Arrays.asList(subEntity1, subEntity2),
 				Collections.singletonList(subEntity4), subEntity3);
 		subEntity1.parent = parentEntity;
@@ -717,6 +720,7 @@ class SubEntity {
 	@Reference
 	SubEntity sibling;
 
+	@Unindexed
 	List<String> stringList;
 
 	String stringProperty;

